### PR TITLE
Some C++ cleanup for header_rewrite

### PR DESCRIPTION
**Pull request by zwoop**

### CODE REVIEW

File: plugins/header_rewrite/conditions.cc [hunk 1]
```diff
@@ -225,8 +223,8 @@ ConditionHeader::append_value(std::string &s, const Resources &res)
     TSDebug(PLUGIN_NAME, "Getting Header: %s, field_loc: %p", _qualifier.c_str(), field_loc);
 
     while (field_loc) {
-      value          = TSMimeHdrFieldValueStringGet(bufp, hdr_loc, field_loc, -1, &len);
-      next_field_loc = TSMimeHdrFieldNextDup(bufp, hdr_loc, field_loc);
+      const char *value = TSMimeHdrFieldValueStringGet(bufp, hdr_loc, field_loc, -1, &len);
+      next_field_loc    = TSMimeHdrFieldNextDup(bufp, hdr_loc, field_loc);
```
Comment (SolidWallOfCode) [inline plugins/header_rewrite/conditions.cc]: Why not move the declaration for `next_field_loc` here as well?

File: plugins/header_rewrite/conditions.h [hunk 2]
```diff
@@ -547,7 +547,7 @@ class ConditionStringLiteral : public Condition
   typedef Matchers<std::string> MatcherType;
 
 public:
-  ConditionStringLiteral(const std::string &v);
+  explicit ConditionStringLiteral(const std::string &v);
```
Comment (SolidWallOfCode) [inline plugins/header_rewrite/conditions.h]: I'd recommend `std::string_view` - it's almost always the better choice for passing in a string. For example, if the call is
```
ConditionStringLiteral("Battle Beast");
```
then a temporary `std::string` will be created, copying the string. For `std::string_view` a temporary will also be made but it won't copy the string.

File: plugins/header_rewrite/operators.h [hunk 3]
```diff
@@ -109,16 +109,17 @@ class OperatorSetRedirect : public Operator
 public:
   OperatorSetRedirect() { TSDebug(PLUGIN_NAME_DBG, "Calling CTOR for OperatorSetRedirect"); }
   void initialize(Parser &p) override;
+
   TSHttpStatus
-  get_status()
+  get_status() const
   {
     return static_cast<TSHttpStatus>(_status.get_int_value());
```
Comment (SolidWallOfCode) [inline plugins/header_rewrite/operators.h]: Heh. You could see if `return { _status.get_int_value() };` works.

File: plugins/header_rewrite/operators.h [hunk 4]
```diff
@@ -109,16 +109,17 @@ class OperatorSetRedirect : public Operator
...
   }
-  std::string
-  get_location(int &size)
+
+  const std::string &
```
Comment (SolidWallOfCode) [inline plugins/header_rewrite/operators.h]: I'd be careful here. This returns a reference to something that could be a temporary, depending on what type `get_value()` returns. If it's really a reference to memory owned by `_location`, I'd return `std::string_view` to make that clear. If you can't return that, you can't return a `std::string` reference either.

File: plugins/header_rewrite/parser.h [hunk 5]
```diff
@@ -66,7 +66,7 @@ class Parser
   }
 
   bool
-  mod_exist(const std::string m) const
+  mod_exist(const std::string &m) const
   {
     return (std::find(_mods.begin(), _mods.end(), m) != _mods.end());
```
Comment (SolidWallOfCode) [inline plugins/header_rewrite/parser.h]: What are those parentheses for? RETURN IS NOT A FUNCTION.

File: plugins/header_rewrite/statement.h [hunk 6]
```diff
@@ -96,38 +96,17 @@ enum NetworkSessionQualifiers {
 class Statement
 {
 public:
-  Statement() : _next(nullptr), _pdata(nullptr), _rsrc(RSRC_NONE), _initialized(false), _hook(TS_HTTP_READ_RESPONSE_HDR_HOOK)
+  Statement() : _next(nullptr), _rsrc(RSRC_NONE), _initialized(false), _hook(TS_HTTP_READ_RESPONSE_HDR_HOOK)
```
Comment (SolidWallOfCode) [inline plugins/header_rewrite/statement.h]: You could move all those to the member declarations.

File: plugins/header_rewrite/conditions.h (hunk 2 above)
Comment (zwoop) [reply plugins/header_rewrite/conditions.h]: Future optimizations.

File: plugins/header_rewrite/parser.h (hunk 5 above)
Comment (zwoop) [reply plugins/header_rewrite/parser.h]: I know, I didn't write this.

File: plugins/header_rewrite/operators.h (hunk 4 above)
Comment (zwoop) [reply plugins/header_rewrite/operators.h]: I'm not changing to string_view in any of this code.

Comment (zwoop) [reply plugins/header_rewrite/operators.h]: Also, the std::string is a member of the Value, and get_value() already returns a reference.

File: plugins/header_rewrite/conditions.cc (hunk 1 above)
Comment (zwoop) [reply plugins/header_rewrite/conditions.cc]: Sure.

File: plugins/header_rewrite/operators.h (hunk 3 above)
Comment (zwoop) [reply plugins/header_rewrite/operators.h]: Didn't compile :-/

``
header_rewrite/operators.h:116:13: error: cannot initialize return object of type 'TSHttpStatus' with an rvalue of type 'int'
    return {_status.get_int_value()};
            ^~~~~~~~~~~~~~~~~~~~~~~
``

**Comment by SolidWallOfCode** [comment]
Check cleared, you're good to go.

**Comment by zwoop** [comment]
Fixed most of the review concerns (but not string_view in this revision). Thanks!

**Comment by bryancall** [comment]
Cherry picked to 8.1.x

**Pull request merged**